src/adhoc.erl

```diff
@@ -105,33 +105,55 @@ produce_response(#adhoc_response{lang = _Lang,
-                                                                    {unit, microsecond}]),
-                        list_to_binary(TS)
-                end,
-    ActionsEls = case Actions of
-                     [] ->
-                         [];
-                     _ ->
-                         ActionsElAttrs = case DefaultAction of
-                                              <<"">> -> [];
-                                              _ -> [{<<"execute">>, DefaultAction}]
-                                          end,
-                         [#xmlel{name = <<"actions">>, attrs = ActionsElAttrs,
-                                 children = [#xmlel{name = Action} || Action <- Actions]}]
-                 end,
-    NotesEls = lists:map(fun({Type, Text}) ->
-                                 #xmlel{name = <<"note">>,
-                                        attrs = [{<<"type">>, Type}],
-                                        children = [#xmlcdata{content = Text}]}
-                         end, Notes),
+    SessionID = ensure_correct_session_id(ProvidedSessionID),
+    ActionsEls = maybe_actions_element(Actions, DefaultAction),
+    NotesEls = lists:map(fun note_to_xmlel/1, Notes),
     #xmlel{name = <<"command">>,
            attrs = [{<<"xmlns">>, ?NS_COMMANDS},
                     {<<"sessionid">>, SessionID},
                     {<<"node">>, Node},
                     {<<"status">>, list_to_binary(atom_to_list(Status))}],
            children = ActionsEls ++ NotesEls ++ Elements}.
+
+-spec ensure_correct_session_id(binary()) -> binary().
+ensure_correct_session_id(SessionID) when is_binary(SessionID), SessionID /= <<"">> ->
+    SessionID;
+ensure_correct_session_id(_) -> 
+    USec = os:system_time(microsecond),
+    TS = calendar:system_time_to_rfc3339(USec, [{offset, "Z"}, {unit, microsecond}]),
+    list_to_binary(TS).
+
+-spec maybe_actions_element([binary()], binary()) -> [exml:element()].
+maybe_actions_element([], _DefaultAction) ->
+    [];
+maybe_actions_element(Actions, <<>>) ->
+    % If the "execute" attribute is absent, it defaults to "next".
+    AllActions = ensure_default_action_present(Actions, <<"next">>),
+    [#xmlel{
+        name = <<"actions">>,
+        children = [#xmlel{name = Action} || Action <- AllActions]
+    }];
+maybe_actions_element(Actions, DefaultAction) ->
+    % A form which has an <actions/> element and an "execute" attribute
+    % which evaluates to an action which is not allowed is invalid.
+    AllActions = ensure_default_action_present(Actions, DefaultAction),
+    [#xmlel{
+        name = <<"actions">>,
+        attrs = [{<<"execute">>, DefaultAction}],
+        children = [#xmlel{name = Action} || Action <- AllActions]
+    }].
+
+-spec ensure_default_action_present([binary()], binary()) -> [exml:element()].
+ensure_default_action_present(Actions, DefaultAction) ->
+    case lists:member(Actions, DefaultAction) of
```

Comment:
Maybe it's possible to cover it in tests?

```suggestion
    case lists:member(DefaultAction, Actions) of
```

Reply:
Committed with `spec` fix.